### DISCUSSION TITLE
case insensitive search

### DISCUSSION THREAD

**Discussion by Docbroke** [leo-arch/clifm#73] [General]
Wouldn't it be better to have `/` search to be case insensitive ?

**Comment by leo-arch** [comment]
That seems like a reasonable request. However, there are a few things involved in the search function (mostly GLOB and REGEX), which I'm not sure are able to perform case insensitive matching, at least by default. Let me see what I can do.

**Reply by leo-arch** [reply]
Done. The quick search function now transforms search expressions _without metacharacters_ into a regular expression instead of a glob expression. So, this: `/do`, is now transformed into this: `.*do.*`, and insofar as regular expressions allow case insensitive matching (unlike glob expressions), search expressions without metacharacters will always be case insensitive from now on.

Searches for expressions containing metacharacters are still performed in a case sensitive manner. I guess I could add a switch to the search function (say `-i`) to allow case insensitive searches. That could work.

**Reply by Docbroke** [reply]
I am getting segmentation fault when I search.

**Reply by leo-arch** [reply]
What are you searching exactly? What's your search expression?

**Reply by Docbroke** [reply]
```
 1  Books_Audio/9         8  Lectures/117         15  bin_arch/302         22  tmp/14               29  tv_channels.m3u
 2  Books_Medical/11      9  Music/8              16  coding/4             23  win-data/            30  user.js
 3  Books_Other/28       10  Originals/81         17  dic/20               24  Examiners.txt        31  waffle_recipe2
 4  Desktop/1            11  Pictures/64          18  dotfiles_alpine/19   25  October.pdf
 5  Documents/76         12  Softwares/36         19  dotfiles_arch/19     26  Scanned Document.pdf
 6  Downloads/86         13  Templates/10         20  mount/17             27  burgerking
 7  Games/16             14  bin/291              21  packages/1           28  mktemp
└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘TIP: Switch workspaces by pressing Alt-[1-4]
Fri, Nov 5, 09:42:14
[1] 09:42 doc:asusalpine ~
<0> $ /music
9 Music
Matches found: 1
Segmentation fault
```
I am getting segfault with any search.

**Reply by leo-arch** [reply]
I'll try to reproduce the crash and fix it.

**Reply by leo-arch** [reply]
Try with the last version (updated a minute ago). I'm not getting any crash.

Btw, I added a new option to the config file: `CaseSensitiveSearch`. It's false by default anyway. This means that all searches (without metacharacters or as regular expressions, will be case insensitive by default).

**Comment by Docbroke** [comment]
still getting segfaults with latest version. I haven't touched config file.

**Reply by leo-arch** [reply]
The config file shouldn't be any problem. I'll keep trying to reproduce the crash.

**Comment by Docbroke** [comment]
I changed config to allow `CaseSensitiveSearch` and now I am getting segfault when search returns true. No segfault when search querry is not found.
```
1  Radio/21                  3  Sudarshan Kriya Short.mp3 5  chime_wm.mp3              7  morning_alarm.mp3
2  bells/29                  4  chime.ogg                 6  chime_wm_edited.mp3       8  twin_bells.wav
└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘Fri, Nov 5, 10:09:07
[1] 10:09 sharad:asusalpine ~/Music
<0> $ /tail
No matches found
Fri, Nov 5, 10:09:12
[1] 10:09 sharad:asusalpine ~/Music
<1> $ /bells
2 bells          8 twin_bells.wav
Matches found: 2
Segmentation fault
```

**Reply by leo-arch** [reply]
Ok. That's a clue.

**Reply by leo-arch** [reply]
I temporarily rolled back changes to make sure it is the new code that is segfaulting.

Please give it a try now.

**Comment by Docbroke** [comment]
No segfaults after role back

**Reply by leo-arch** [reply]
Great. At least I know now what piece of code is segfaulting. Though, since I'm still not able to reproduce the crash, I'm working kinda blindly.

**Reply by leo-arch** [reply]
Could you tell me what's the value of Alpine Linux default CFLAGS? Maybe it has something to do with the way compilation is done. Maybe.

**Reply by Docbroke** [reply]
I am not sure. I guess this may help, this is the line from output when I compile clifm using "abuild -r". Alpine uses APKBUILD (similar to PKGBUILD of arch), and compilation is done using abuild (like makepkg).
```
gcc -o clifm src/*.c -Os -fomit-frame-pointer -Wl,--as-needed,-O1,--sort-common -lreadline -lacl -lcap -lmagic
```
This is APKBUILD I am using
```
pkgname=clifm
pkgver=1.2.2_git20211105
pkgrel=4
_commit="0695883ce287490c1d2ba6f0fdb3d47b5f120fb2"
pkgdesc="CLI file manager"
url="https://github.com/leo-arch/clifm"
arch="all"
license="GPL-2.0-or-later"
makedepends="acl-dev file-dev gettext-tiny-dev libcap-dev readline-dev"
options="!check" # no test suite
#subpackages="$pkgname-doc $pkgname-bash-completion $pkgname-zsh-completion"
subpackages="$pkgname-doc $pkgname-bash-completion"
source=$pkgname-$_commit.tar.gz::"https://github.com/leo-arch/clifm/archive/$_commit/clifm-$_commit.tar.gz"
builddir="$srcdir/$pkgname-$_commit"

build() {
	make
}

package() {
	make PREFIX=/usr MANDIR=/usr/share/man DESTDIR="$pkgdir" install
}

sha512sums="
d6f7ad24d8eaebad4251ebe2e680debcaf809afadea2df55cdd38300bb9c9dd83e9d83cd084924002077bde58ce9975ccfda2bdcd78d025c191624af4f772e67  clifm-0695883ce287490c1d2ba6f0fdb3d47b5f120fb2.tar.gz
"
```

**Reply by leo-arch** [reply]
I've tried using those exact compilation parameters. Still no crash (not even a mem leak in Valgrind). I guess I should try on Alpine itself: maybe something related to musl libc.

**Reply by leo-arch** [reply]
Yeah, got it segfaulting. Now I can fix it.

EDIT: Fixed. It was just a double call to `regfree(3)`. This has been there for a while, but for some reason it started crashing just now, and only on Alpine (or musl systems).

Btw, I liked Alpine (never tried it before): really simple and minimalist. and `apk` is quite fast.

**Reply by Docbroke** [reply]
That's incredible, you installed alpine, compiled clifm and fixed it in 2 hours.
Alpine is nice, minimal & systemd free. I moved to alpine after using arch for about 4-5 years.

**Reply by leo-arch** [reply]
Strictly speaking, I didn't install Alpine, but just run the live medium on qemu, enabled networking, added repos, and installed the packages I needed to compile and fix CliFM. Besides, the problem was much easier to solve than I expected.

If I were not so used to Arch (6 years using it as my only driver), I'd move to Alpine without thinking: there's no chance for an Arch user to dislike Alpine. However, my system is not that far from Alpine: OpenRC, a hardened kernel, and a really minimalist setup.
